### DISCUSSION TITLE
Correctly get select line range when using command and key mappings?

### DISCUSSION THREAD

**Discussion by linrongbin16** [neovim/neovim#26092] [Q&A]
There are two methods to get the select line range in neovim now: use command or key mappings.

here is an example:

first define the command that support range select:

```lua
vim.api.nvim_create_user_command('MyCommand', function(opts)
  -- method-1. get range from user commamd
  local lstart1 = opts.line1
  local lend1 = opts.line2

  -- method-2. get range from vim function
  local m = vim.fn.mode() -- detect current mode
  if m == 'v' or m == 'V' or m == '\22' then -- <C-V>
    vim.cmd([[execute "normal! \<ESC>"]])
    local lstart2 = vim.fn.getpos("'<")[2]
    local lend2 = vim.fn.getpos("'>")[2]
  end

  -- method-3. another vim function to detect visual mode
  -- the behavior is quite different, it always get the last visual select line range, so the result can be wrong.
  local m2 = vim.fn.visualmode()
  local lstart3 = vim.fn.getpos("'<")[2]
  local lend3 = vim.fn.getpos("'>")[2]
end, { range = true })
```

define key mappings on the command:

```lua

-- key-1. use `<cmd>`
vim.keymap.set(
  {"n", "x"},
  "<leader>ga",
  "<cmd>MyCommand<cr>",
  { noremap = true }
)

-- key-2. use `:`
vim.keymap.set(
  {"n", "x"},
  "<leader>gb",
  ":MyCommand<CR>",
  { noremap = true }
)
```

Now, I have 3 ways to select the lines and execute the command:

1. (select lines and) manually type `<,>:MyCommand` and type `enter`:
   - the `opts.line1` and `opts.line2` (method-1) could work correctly.
   - the `vim.fn.mode()` (method-2) not work, it doesn't treat it as visual mode.
   - the `vim.fn.visualmode()` (method-3) works, it's the same with method-1.
2. (select lines and) press `<leader>ga`(key-1):
   - the `opts.line1` and `opts.line2` (method-1) don't work, it doesn't recognize the line range.
   - the `vim.fn.mode()` (method-2) works correctly, it catched the visual mode, and use `execute "normal! \<ESC>"` to first quit visual mode, go back to normal, then use `getpos()` to get the line range.
3. (select lines and) press `<leader>gb` (key-2), the result is same with the 1st one (e.g. manual enter command).

my question is, as a plugin writer, how should I provide a user command, so make sure above 3 use cases working correctly, to support line range?

**Comment by linrongbin16** [comment]
I have written two plugins, both need to detect the visual selection.

## Use-case-1

In the [gitlinker.nvim's GitLink](https://github.com/linrongbin16/gitlinker.nvim/blob/cafacd0bbc9c6f5e89d4e5366d64a7ee9f239c98/lua/gitlinker.lua#L411-L428), I want the command `GitLink` work for both normal mode and visual mode.

In normal mode, it select the current line on cursor. In visual mode, it select the line ranges.

For now I just merge two results from both `vim.fn.getpos()` and `opts.line1`+`opts.line2`, so the 3 cases can be handled. but somehow I feel it's not elegant.


## Use-case-2

In the [fzfx.nvim's FzfxLiveGrepV](https://github.com/linrongbin16/fzfx.nvim/blob/main/lua/fzfx/fzf_helpers.lua#L76-L82), I need the command `FzfxLiveGrepV` to get user selection.

But I only allow user to mapping this command in visual mode (there's a technical limitation, I will discuss it later).

So the implementation is like:

```lua
vim.cmd([[ execute "normal! \<ESC>" ]])
local mode = vim.fn.visualmode()
local start_pos = vim.fn.getpos("'<")
local end_pos = vim.fn.getpos("'>")
```

since user is required to only mapping it to visual mode, so it always first execute `normal! \<esc>` to go back to normal mode, then use `getpos()` to get the selection range. which is always correct.

But just like I mentioned before, this command only allow user to map to visual mode. I had to provide another command `FzfxLiveGrep` for user to use in normal mode. the `FzfxLiveGrep` could allow user to input some arguments as input info.

I wish I could merge these two commands into one, which seems much more neat and clean.

Is there any optimization to do?

**Comment by tkapous** [comment]
I faced a similar situation where I need to get the range from a xmap, and get back to visual selection.
Problem is, in linewise-visual, `getpos()` will return a 'large number' (aka int max, from my tests):

`:h getpos()`
> Note that for '< and '> Visual mode matters: when it is "V"
> (visual line mode) the column of '< is zero and the column of
> '> is a large number equal to |v:maxcol|.

However, `getcharpos()` returns the expected number (the last column in the line):

```lua
vim.cmd([[ execute "normal! \<ESC>" ]])
local start_pos = vim.fn.getcharpos("'<")
local end_pos = vim.fn.getcharpos("'>")
vim.cmd([[ execute "normal! gv" ]])
```
